Cleanup attr.def

davezarzycki opened the pull request:
This builds on #16014, so please ignore the first commit.

1) Formalize `OnAccessor`. A hack used to alias this to `OnFunc`.
2) New aggregates: `OnNominalType`, `OnGenericType`, `OnAbstractFunction`.
3) Consistent and self-documented (albeit custom) style in Attr.def to ease code review/audits.

@rjmccall and @slavapestov – Review questions:

1) Now that `OnAccessor` is formalized, should it be removed from any attributes?
2) Now that `OnAccessor` is formalized, should `OnFunc` be removed from any attributes?
3) Now that new aggregate "OnXYZ" options exist, it seems like there might be some oversights in the `.def` file. For example, it is intentional that `inline` doesn't work on destructors? Also, why doesn't `_alignment` include `OnClass`? There might be more oversights. Feedback would be appreciated.
4) The following seems to be a common attribute requirement: `OnAbstractFunction & ~OnDestructor`. Should it be formalized? Perhaps `OnNonDestructorFunctions`?

davezarzycki committed:
[AST] NFC: Attr.def clean up

1) Formalize "OnAccessor". A hack used to alias this to "OnFunc".
2) New aggregates: OnNominalType, OnGenericType, OnAbstractFunction.
3) Consistent and self-documented (albeit custom) style to ease code review/audits.
4) Removes a few cases of `OnAccessor` based on pull request #16031 feedback.

jrose-apple commented:
Classes can't be realigned today. Destructors…could be inlinable, I guess, but that would only ever come into play if you knew a class instance didn't escape, which seems super rare across module boundaries (since it's valid for an initializer to escape `self`).

slavapestov commented:
`@inline(never)` and `@inline(always)` don't work on destructors because I don't think there's much value in this. The stdlib does define a few `@inlinable` destructors though.

I'm not a fan of the idea of `OnNonDestructorFunctions`. I'd rather we spell it out explicitly.

slavapestov commented:
@davezarzycki I'm not sure I understand your question about inlining.

davidungar commented:
@davezarzycki I'm not sure why I'm a reviewer on this PR. It doesn't seem to be code I've been into. What sort of input from me would be helpful?

davezarzycki commented:
Hi @slavapestov, @rjmccall, and @jrose-apple,

Just to be clear here, I want to know whether the new "OnXYZ" flags in this pull request expose any clean up opportunities or oversights in the existing `.def` file. If it doesn't, then I'd like to seek approval for this pull request.

I'm primarily interested in questions 1 and 2 at the top of this pull request: does the formalization of `OnAccessor` exposes any clean up opportunities? Any feedback would be appreciated.

I'm less concerned about questions 3 and 4. I just wanted to double check with people that there weren't any oversights with the existing "OnXYZ" flags while I was cleaning up the file.

Finally, I do think that aggregate "OnXYZ" flags are incredibly valuable, especially for people less familiar with the code, or people trying to extend the code. For example, `OnNominalType` is both more clear and more future proof than itemizing the current list of nominal types over multiple attributes. Toward that end, if any other useful aggregate flag combinations exist, I'd be happy to look into adding them. I suggested "on non-destructor functions" because it seemed like both a valid intent and a common pattern. I think other aggregates are potentially useful too.

PS –  Hi @davidungar. I misinterpreted how GitHub reports "reviewers" in their user interface. I thought somebody requested you as a reviewer in #16014 and therefore you should be added to this pull request too. Upon closer inspection, it seems that GitHub reports any *commenter* on pull request code as a "reviewer", which is confusing.

davezarzycki commented:
Ping. Any feedback/review on this pull request would be appreciated.

rjmccall commented on include/swift/AST/Attr.def:
```diff
@@ -98,255 +98,281 @@ TYPE_ATTR(thick)
+CONTEXTUAL_SIMPLE_DECL_ATTR(postfix, Postfix,
+  OnFunc | OnAccessor | OnOperator |
+  DeclModifier,
+  25)
```
infix/prefix/postfix can't ever be on accessors.

rjmccall commented on include/swift/AST/Attr.def:
```diff
@@ -98,255 +98,281 @@ TYPE_ATTR(thick)
+SIMPLE_DECL_ATTR(LLDBDebuggerFunction, LLDBDebuggerFunction,
+  OnFunc | OnAccessor |
+  UserInaccessible,
+  17)
```
I don't think this can be on accessors.

rjmccall commented on include/swift/AST/Attr.def:
```diff
@@ -98,255 +98,281 @@ TYPE_ATTR(thick)
-                 OnFunc, 11)
-
+  OnFunc | OnAccessor,
+  11)
```
I don't think `IBAction` can be on accessors.

davezarzycki commented:
Hi @rjmccall – I've addressed your feedback and added tests. For whatever it may be worth, the operator attributes don't even parse in a accessor context.

@swift-ci please smoke test

rjmccall commented:
Yeah, those attributes require very specific syntax which is tied in the parser to func declarations.
